docs/reusing-typecheck-results.md

```diff
@@ -0,0 +1,65 @@
+---
+title: Reusing typecheck results
+category: Compiler Internals
+categoryindex: 200
+index: 42
+---
+
+# Reusing typecheck results between compiler and tooling runs
+
+Caching and reusing typecheck results between compiler and tooling runs will be an optimization technique aimed at improving the performance and efficiency of the F# development experience.
+
+### Why is it important
+
+- **Performance**. Recomputing type information for large projects is time-consuming. By caching the results, subsequent runs can skip this step, leading to faster builds.
+
+- **IDE responsiveness**. IDEs can provide faster IntelliSense, code navigation, and other features if they can quickly access cached type information.
+
+- **Error reduction**. Consistent typechecking results help avoid scenarios where slight differences between runs could cause different errors or warnings to be produced.
+
+### General considerations
+
+- The technique will be implemented separately for the compiler and tooling, for better granularity and trackability.
```

Comment:
I would rather twist the argumentation around.
"The feature will get the best benefits if data is reused between tooling and compiler runs. However, for practical reasons (versioning, project settings, runtime difference) it will not be implemented from the start".

i.e. the separation is rather a technical limitation at the moment, but should not become part of the axioms that affect the overall design.

If those limits are lifted (e.g. tooling using the same F.C.S. as building), the potential is even bigger - running tests or starting a debugging session could be utilizing a fully cached typecheck result (and consequently imports, parsing,..)

Reply:
Agree. I decided to not focus on tooling for now since there is a lot to decide on even in the compiler. I put the tooling support to the followups.

Reply:
I did not mean the area of biggest potential as supporting tooling at all, it was meant as supporting compilation (build, test, run, debug,..) with information collected **from** tooling.

That way, starting tests could feel almost instantaneous, because the type checking would be already happening while user is making changes.